### PR TITLE
1594 Using ECR as an external registry

### LINKED ISSUE #1594: Use ECR as an external registry during 'Zarf Init'
### Is your feature request related to a problem? Please describe.
**As** Ashton **I want** an easy way to deploy Zarf configured with ECR as an external registry **so that** I can take advantage of the HA benefits of running a registry out of the cluster in AWS.

### Describe the solution you'd like

- **Given** I have a fresh EKS cluster and IAM access to ECR
- **When** I run `zarf init` with an ECR registry domain configured
- **Then** Zarf successfully initializes the cluster
- **And** Creates the ECR repositories for me
- **And** _(optional for this issue)_ Creates a mechanism to keep the `private-registry` credentials up to date.

### Additional context
ECR has a few 'unique' quirks that make it difficult to use as an external registry at the moment.

1. ECR credentials are only valid for 12 hours, so something within clusters would have to refresh the zarf image pull secrets so that the kubelet is able to authenticate and pull images from ECR during package deployments.

2. ECR registries do not have the ability to 'push to create' repositories that do not exist yet. Since Zarf packages will be pushing all sorts of images into the registry when the package is being deployed, this is a pretty big blocker.

Using ECR as an external registry would be really useful for a lot of people so we should implement some solution to the two issues above.

### PR DESCRIPTION
**Pull request by YrrepNoj**
## Description
This PR enables users who are using an EKS initialized Zarf cluster to use ECR as their external registry.

...

## Related Issue

Fixes #1594 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (security config, docs update, etc)

## Checklist before merging

- [ ] Test, docs, adr added or updated as needed
- [ ] [Contributor Guide Steps](https://github.com/defenseunicorns/zarf/blob/main/CONTRIBUTING.md#developer-workflow) followed

### CODE REVIEW

File: src/internal/packager/hook/ecr.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+	if err != nil {
+		return err
+	}
+
+	// Get the username and password from the auth token
+	data, err := b64.StdEncoding.DecodeString(authToken)
+	if err != nil {
+		return fmt.Errorf("unable to decode the ECR authorization token: %w", err)
+	}
+	username := "AWS"
```
Comment (Noxsios) [inline src/internal/packager/hook/ecr.go:49]: const

File: src/internal/packager/hook/ecr.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+		return "", fmt.Errorf("unable to get the ECR authorization token: %w", err)
+	}
+
+	return *authToken.AuthorizationData.AuthorizationToken, nil
+}
+
+func fetchAuthToPrivateECR(registryURL string, region string) (string, error) {
+	ecrClient := ecr.New(session.New(&aws.Config{Region: aws.String(region)}))
+	authToken, err := ecrClient.GetAuthorizationToken(&ecr.GetAuthorizationTokenInput{})
+	if err != nil || len(authToken.AuthorizationData) == 0 || authToken.AuthorizationData[0] == nil {
```
Comment (Noxsios) [inline src/internal/packager/hook/ecr.go:85]: comment the differences between these two functions

File: src/internal/packager/hook/ecr.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+// CreateTheECRRepos creates an ecr repository for each image provided
+func CreateTheECRRepos(ecrHook types.HookConfig, images []string) error {
+	registryPrefix := ecrHook.HookData["repositoryPrefix"].(string)
+	region := ecrHook.HookData["region"].(string)
+	registryURL := ecrHook.HookData["registryURL"].(string)
+
+	if registryPrefix != "" {
+		registryPrefix += "/"
+	}
```
Comment (Noxsios) [inline src/internal/packager/hook/ecr.go:100]: better check here pls 😉

File: src/internal/packager/hook/ecr.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+		return errors.New("no docker config file found, run 'zarf tools registry login --help'")
+	}
+
+	// Save the credentials to the docker.config file
+	configs := []*configfile.ConfigFile{cfg}
+	authConfig := docker_types.AuthConfig{Username: username, Password: password, ServerAddress: registryURL}
+	err = configs[0].GetCredentialsStore(registryURL).Store(authConfig)
+	if err != nil {
+		return fmt.Errorf("unable to get credentials for %s: %w", registryURL, err)
+	}
```
Comment (Noxsios) [inline src/internal/packager/hook/ecr.go:67]: throw this in `utils/image.go` 😉

File: src/internal/packager/images/push.go [hunk 5]
```diff
@@ -53,7 +54,9 @@ func (i *ImgConfig) PushToZarfRegistry() error {
 	craneTransport := utils.NewTransport(httpTransport, progressBar)
 
 	pushOptions := config.GetCraneOptions(i.Insecure, i.Architectures...)
-	pushOptions = append(pushOptions, config.GetCraneAuthOption(i.RegInfo.PushUsername, i.RegInfo.PushPassword))
+	if i.RegInfo.RegistryType != types.ECRRegistry {
```
Comment (Noxsios) [inline src/internal/packager/images/push.go:57]: comment here pls

File: src/pkg/packager/common.go [hunk 6]
```diff
@@ -30,11 +30,12 @@ import (
...
-	cluster  *cluster.Cluster
-	tmp      types.TempPaths
-	arch     string
-	warnings []string
+	cfg         *types.PackagerConfig
+	cluster     *cluster.Cluster
+	tmp         types.TempPaths
+	arch        string
+	warnings    []string
+	pluginHooks map[string]types.HookConfig
```
Comment (Noxsios) [inline src/pkg/packager/common.go:38]:
```suggestion
	hooks map[string]types.HookConfig
```

File: src/pkg/packager/common.go [hunk 7]
```diff
@@ -60,7 +61,8 @@ func New(cfg *types.PackagerConfig) (*Packager, error) {
 	var (
 		err       error
 		pkgConfig = &Packager{
-			cfg: cfg,
+			cfg:         cfg,
+			pluginHooks: make(map[string]types.HookConfig),
```
Comment (Noxsios) [inline src/pkg/packager/common.go:65]:
```suggestion
			hooks: make(map[string]types.HookConfig),
```

File: src/pkg/packager/hooks.go [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+	for _, hookSecret := range hookSecrets.Items {
+		hookConfig := types.HookConfig{}
+
+		// Get any data from the hook secret
+		err := json.Unmarshal(hookSecret.Data["data"], &hookConfig)
+		if err != nil {
+			return fmt.Errorf("unable to load the hook configuration for the %s hook: %w", hookSecret.Name, err)
+		}
+
+		p.pluginHooks[hookSecret.Name] = hookConfig
```
Comment (Noxsios) [inline src/pkg/packager/hooks.go:33]:
```suggestion
		p.hooks[hookSecret.Name] = hookConfig
```

File: src/pkg/packager/hooks.go [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+func (p *Packager) runPostComponentHooks(component types.ZarfComponent) error {
+	return p.runComponentHooks(types.AfterComponent, component)
+}
+
+func (p *Packager) runPackageHooks(lifecycle types.HookLifecycle) error {
+
+	// If we are not able to run the hooks, return early
+	if p.cluster == nil || p.pluginHooks == nil {
```
Comment (Noxsios) [inline src/pkg/packager/hooks.go:58]:
```suggestion
	if p.cluster == nil || p.hooks == nil {
```

File: src/pkg/packager/hooks.go [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+		return nil
+	}
+
+	for _, hookConfig := range p.pluginHooks {
```
Comment (Noxsios) [inline src/pkg/packager/hooks.go:62]:
```suggestion
	for _, hookConfig := range p.hooks {
```

File: src/pkg/packager/deploy.go [hunk 11]
```diff
@@ -82,13 +83,45 @@ func (p *Packager) Deploy() error {
 
 	// Filter out components that are not compatible with this system
 	p.filterComponents(true)
+	componentsToDeploy := p.getValidComponents()
+
+	// Check if any of the components we are going to deploy require a cluster
+	requiresCluster := false
+	for _, component := range componentsToDeploy {
+		requiresCluster = requiresCluster || component.RequiresCluster()
```
Comment (Noxsios) [inline src/pkg/packager/deploy.go:91]: make this clean 😉

File: src/pkg/packager/deploy.go [hunk 12]
```diff
@@ -82,13 +83,45 @@ func (p *Packager) Deploy() error {
...
+	// NOTE: We are not doing this for init packages at this time, init packages will manage hooks on their own.
+	if requiresCluster && !p.cfg.IsInitConfig {
+		// Make sure we have access to the cluster
+		var err error
+		if p.cluster == nil {
+			p.cluster, err = cluster.NewClusterWithWait(cluster.DefaultTimeout, true)
+			if err != nil {
+				return fmt.Errorf("unable to connect to the Kubernetes cluster: %w", err)
+			}
+		}
```
Comment (Noxsios) [inline src/pkg/packager/deploy.go:104]: you are your broken code 😉

File: src/pkg/packager/deploy.go [hunk 13]
```diff
@@ -82,13 +83,45 @@ func (p *Packager) Deploy() error {
...
+			return fmt.Errorf("unable to get all cluster hooks: %w", err)
+		}
+
+		if err := p.runPreDeployHooks(); err != nil {
+			return fmt.Errorf("unable to run pre-deploy hooks: %w", err)
+		}
+	}
+
+	// Deploy all of the components
+	deployedComponents, err := p.deployComponents(componentsToDeploy)
```
Comment (Racer159) [inline src/pkg/packager/deploy.go:116]: We may want to drop "package" hooks and go with "runonce" component hooks instead

File: src/pkg/packager/deploy.go [hunk 14]
```diff
@@ -167,6 +199,31 @@ func (p *Packager) deployInitComponent(component types.ZarfComponent) (charts []
...
+					message.Warnf("Unable to authenticate to ECR: %s", err.Error())
+				}
+
+				if err := hook.CreateTheECRRepos(hookConfig, component.Images); err != nil {
+					return charts, fmt.Errorf("unable to create the ECR repositories: %w", err)
+				}
+
+				p.cfg.InitOpts.RegistryInfo.RegistryType = types.ECRRegistry
+			}
+		}
```
Comment (Racer159) [inline src/pkg/packager/deploy.go:225]: This should run all hooks not just ECR

**Review by Racer159 — CHANGES_REQUESTED** [review]

**Review by Noxsios — CHANGES_REQUESTED** [review]
some changes 😉

File: src/internal/packager/hook/ecr.go (hunk 2 above)
Comment (jeff-mccoy) [reply src/internal/packager/hook/ecr.go:85]: lint things

File: src/internal/packager/hook/ecr.go [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+	for _, image := range images {
+		// Parse the image ref
+		imageRef, err := transform.ParseImageRef(image)
+		if err != nil {
+			return fmt.Errorf("unable to parse the image ref: %w", err)
+		}
+
+		repositoryName := registryPrefix + imageRef.Path
+		if strings.Contains(registryURL, publicECRRegistryURL) {
+			_, err = ecrPublicClient.CreateRepository(&ecrpublic.CreateRepositoryInput{RepositoryName: aws.String(repositoryName)})
```
Comment (jeff-mccoy) [inline src/internal/packager/hook/ecr.go:120]: these lines are really long 😆 is there no line length enforcement happening in this repo anymore?

**Review by jeff-mccoy — COMMENTED** [review]
Others picked up most of the code change details already. More broadly I have concerns as this is a pretty large lifecycle change / addition with 

- no ADR or design explanation I can find
- no doc updates 

It feels like a mashup of extensions and actions, and adds a lot of entry points. Really feel like this needs a lot more in terms of the "why" and future plan here vs just "here's some code".

File: src/pkg/packager/deploy.go (hunk 13 above)
Comment (YrrepNoj) [reply src/pkg/packager/deploy.go:118]: I am not going to do this right now, I am leaving a TODO comment for this and I"ll be sure to discuss the different approaches in the ADR that we are working on.

**Comment by YrrepNoj** [comment]
Closing as it appears we are not going for the webhook approach instead <3

**Pull request closed without merging**